runtime/src/qio/regex/bundled/re2-interface.cc

```diff
@@ -320,32 +319,194 @@ qio_bool qio_regex_match(qio_regex_t* regex, const char* text, int64_t text_len,
+  buf_len += data_len;
+
+  return true; // everything is OK
+}
+
+static bool append_rewrite(char*& buf, // buffer
+                           int64_t& buf_sz, // allocated size
+                           int64_t& buf_len, // amount used
+                           const StringPiece& rewrite,
+                           const StringPiece* vec,
+                           int veclen) {
+  for (const char *s = rewrite.data(), *end = s + rewrite.size();
+       s < end;
+       s++) {
+    if (*s != '\\') {
+      // append 1 byte
+      append(buf, buf_sz, buf_len, s, 1);
+      continue;
+    }
+    s++;
+    int c = (s < end) ? *s : -1;
+    if (isdigit(c)) {
+      int n = (c - '0');
+      if (n >= veclen) {
+        return false;
+      }
+      StringPiece snip = vec[n];
+      if (!snip.empty())
+        append(buf, buf_sz, buf_len, snip.data(), snip.size());
+    } else if (c == '\\') {
+      const char* backslash = "\\";
+      append(buf, buf_sz, buf_len, backslash, 1);
     } else {
-      if( RE2::Replace(&s, *re, rewrite) ) {
-        ret = 1;
-      } else {
-        ret = 0;
+      return false;
+    }
+  }
+  return true;
+}
+
+static int free_and_return_error(char* buf) {
+  if (buf != NULL) qio_free((void*)buf);
+  return -1;
+}
+
+// these should match re2.cc
+static const int kMaxArgs = 16;
+static const int kVecSize = 1+kMaxArgs;
+
+// this function is loosly based upon RE2::GlobalReplace
+static int replace(const RE2& re, const StringPiece& rewrite,
+                   const StringPiece& str,
+                   int maxreplace,
+                   const char** str_out, int64_t* len_out) {
+  StringPiece vec[kVecSize];
+  int nvec = 1 + RE2::MaxSubmatch(rewrite);
+  if (nvec > 1 + re.NumberOfCapturingGroups())
+    return -1;
+  if (nvec > static_cast<int>(sizeof(vec)/sizeof(StringPiece)))
+    return -1;
+
+  const char* p = str.data();
+  const char* ep = p + str.size();
+  const char* lastend = NULL;
+
+  char* buf = NULL;
+  int64_t buf_sz = 0;
+  int64_t buf_len = 0;
+
+  int count = 0;
+  while (p <= ep) {
+    // don't try to replace anything if we are replacing 0
+    if (maxreplace == 0)
+      break;
+
+    if (!re.Match(str, static_cast<size_t>(p - str.data()),
+                  str.size(), RE2::UNANCHORED, vec, nvec))
+      break;
+
+    // append the data before the match
+    if (p < vec[0].data())
+      if (!append(buf, buf_sz, buf_len, p, vec[0].data() - p))
+        return free_and_return_error(buf);
+
+    if (vec[0].data() == lastend && vec[0].empty() && count > 0) {
+      // quit if we are at the end (trailing \0 added later)
+      if (p == ep) {
+        break;
+      }
+
+      // Disallow empty match at end of last match: skip ahead.
+      //
+      if (re.options().encoding() == RE2::Options::EncodingUTF8) {
+        // append the character
+        int bad;
+        int nbytes = 0;
+        int32_t chr = 0;
+        bad = chpl_enc_decode_char_buf_utf8(&chr, &nbytes, p, ep - p, false);
```

Comment:
A couple of tiny nitpicks:

1. Why declare `bad` a few lines above here?
2. Could `bad` be called `err` or `ret`?